### PR TITLE
feat: [CDS-53545]: v1 EOL banner APIs

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
1. Include version info in service response. Hide the request/response param from openapi 
2. Introduce 2 new APIs for checking if we want to show the banner for EOL: pipeline and template
3. **/template** handles showing banner on Temp studio for both pipeline and stage template
4. **/pipeline** handles showing banner on pipeline studio. 
5. Along with banner return list of stage ids that are of v1 structure.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@Api("/eol-banner")
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java:39]: lets use service v1 prefix here

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+  @Inject private TemplateResourceClient templateResourceClient;
+
+  public EOLBannerResponseDTO checkPipelineUsingV1Stage(
+      @NonNull String accountId, @NonNull EOLBannerRequestDTO requestDTO) {
+    boolean showBanner = false;
+    List<String> stageIdentifiersUsingV1Stage = new ArrayList<>();
+
+    final PMSPipelineResponseDTO existingPipeline =
+        NGRestUtils.getResponse(pipelineServiceClient.getPipelineByIdentifier(requestDTO.getPipelineIdentifier(),
+            accountId, requestDTO.getOrgIdentifier(), requestDTO.getProjectIdentifier(), null, null, null));
```
[the remark this answers is not in the document]
Comment (yogesh-chauhan) [reply 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:71]: Lets update pipeline service client

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    if (TemplateRefHelper.hasTemplateRef(pipelineYaml)) {
+      try {
+        // not checking for template access
+        pipelineYaml =
+            NGRestUtils
+                .getResponse(templateResourceClient.applyTemplatesOnGivenYamlV2(accountId,
+                    requestDTO.getOrgIdentifier(), requestDTO.getProjectIdentifier(), null, null, null, null, null,
+                    null, null, null, null,
+                    TemplateApplyRequestDTO.builder().originalEntityYaml(pipelineYaml).checkForAccess(false).build(),
+                    false))
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:90]: lets check for cache here

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                .getMergedPipelineYaml();
+      } catch (Exception ex) {
+        throw new InvalidRequestException(
+            format("error occurred in generating resolved pipeline yaml: %s", ex.getMessage()));
+      }
+      pipelineYamlField = getYamlField(pipelineYaml, "pipeline");
+    }
+    ArrayNode stageArrayNode = (ArrayNode) pipelineYamlField.getNode().getField("stages").getNode().getCurrJsonNode();
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:98]: can we check for null getField("stages")

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+  private YamlField getYamlField(String yaml, String fieldName) {
+    try {
+      return YamlUtils.readTree(yaml).getNode().getField(fieldName);
+    } catch (Exception e) {
+      throw new InvalidRequestException(
+          format("not able to parse %s yaml because of error: %s", fieldName, e.getMessage()));
+    }
+  }
+
+  private boolean checkStageNode(@NonNull EOLBannerRequestDTO requestDto, @NonNull String accountId,
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:139]: can we call it isV1Stage ?

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+  }
+
+  private boolean checkStageWithTemplate(
+      YamlNode stageNode, String accountId, EOLBannerRequestDTO requestDto, List<String> stageIdentifiers) {
+    String stageYaml = YamlPipelineUtils.writeYamlString(stageNode.getCurrJsonNode());
+    String resolvedStageYaml =
+        NGRestUtils
+            .getResponse(templateResourceClient.applyTemplatesOnGivenYamlV2(accountId, requestDto.getOrgIdentifier(),
+                requestDto.getProjectIdentifier(), null, null, null, null, null, null, null, null, null,
+                TemplateApplyRequestDTO.builder().originalEntityYaml(stageYaml).checkForAccess(false).build(), false))
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:169]: cache

File: 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+  }
+
+  private boolean checkDeploymentStageConfigIsV1(DeploymentStageConfig deploymentStageConfig) {
+    return deploymentStageConfig.getServiceConfig() != null || deploymentStageConfig.getInfrastructure() != null;
+  }
+
+  public boolean checkStage(
+      YamlNode stageNode, String accountId, EOLBannerRequestDTO requestDto, List<String> stageIdentifiers) {
+    String stageYaml = YamlPipelineUtils.writeYamlString(stageNode.getCurrJsonNode());
+    DeploymentStageConfig deploymentStageConfig = getDeploymentStageConfig(stageYaml);
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java:188]: can we just use getNode or getField here to decide v1

File: 960-ng-core-beans/src/main/java/io/harness/ng/core/service/dto/ServiceResponseDTO.java [hunk 8]
```diff
@@ -34,4 +35,5 @@ public class ServiceResponseDTO {
   Map<String, String> tags;
   @JsonIgnore Long version;
   String yaml;
+  @Schema(hidden = true) boolean newVersion;
```
Comment (yogesh-chauhan) [inline 960-ng-core-beans/src/main/java/io/harness/ng/core/service/dto/ServiceResponseDTO.java:38]: lets just call it v2Service ?

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)* status: 11 file(s) to review. 1 change requests. 9. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 71 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-71:-NPS1gqnDQjmtRASYysK:bx84fui) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L71)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Lets update pipeline service client
</blockquote></details>

Done. Tested w/ remote pipelines, pipeline templates and stage templates 
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-90:-NPS1mZg6RlcdCnhBQ7w:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L90)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets check for cache here
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 98 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-98:-NPS1oqe0n6HxwQT8iWH:bbjuibk) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L98)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we check for null getField("stages")
</blockquote></details>

All pipelines need to have stages. We're good here. I've added null checks where reqd.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 139 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-139:-NPS1vrzA2V1SUl1B9DM:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L139)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we call it isV1Stage ?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 169 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-169:-NPS1wNf4XcjfdAkwCe8:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L169)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

cache
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pb-r1-188:-NPS1xSq6MVi3ePIjazl:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/EOLBannerService.java#L188)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we just use getNode or getField here to decide v1
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java` line 39 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pe-r1-39:-NPS1zlE4SVVA9_LnSBf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java#L39)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets use service v1 prefix here
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java` line 49 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NPH1QUd94yz9q0tJ4pe-r1-49:-NPS2-0N-inY4f-Ju4F-:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/120-ng-manager/src/main/java/io/harness/ng/core/eolbanner/resources/EOLBannerResource.java#L49)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets hide this resource
</blockquote></details>

Done.
___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/service/dto/ServiceResponseDTO.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/44368#-NP8jcK_d9YJ0lLEeE_Z-r1-38:-NPS20g77rZUF4Qc9TPs:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/debeda3a881144543f4272ac13f6661ba590aeaf/960-ng-core-beans/src/main/java/io/harness/ng/core/service/dto/ServiceResponseDTO.java#L38)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets just call it v2Service ?
</blockquote></details>

Done.

**Comment by yogesh-chauhan** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/44368#-NPSL46J7qSRr3QGWJNg:-NPSL46K5D4Aexkf_5Qp:bpgm0e8) ([raw file](https://github.com/harness/harness-core/blob/022c32fc96de0cb7312dd0a602dcf3534a8a1a2d/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java#L76)):*
> ```Java
>         serviceResponseDTO.setV2Service(true);
>       }
>     }
> ```

can we do this
```
if (includeVersionInfo && serviceEntity.getType() != null) {
      serviceResponseDTO.setV2Service(true);
    }
```

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 2 of 8 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)* status: 11 file(s) to review. 0 of 2 approvals obtained. 

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/44368#-NPSL46J7qSRr3QGWJNg:-NPT2pbOBGVpTv0nnw03:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/022c32fc96de0cb7312dd0a602dcf3534a8a1a2d/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java#L76)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we do this
```
if (includeVersionInfo && serviceEntity.getType() != null) {
      serviceResponseDTO.setV2Service(true);
    }
```
</blockquote></details>

Done

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 2 of 8 files at r1, 7 of 9 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44368)* status: :shipit: complete! ✓

**Pull request merged**